### PR TITLE
Peep order

### PR DESCRIPTION
**Pull request by Broxzier**
This PR will sort the guest/staff lists while taking numbers into account. Numbers in the middle of the string will be used for sorting too, an example screenshot:

Staff list

The second commit will sort the peep list when a park is loaded. It's a bit hacky, because the function `peep_update_name_sort` assumes the list is in order already.

### CODE REVIEW

**Comment by IntelOrca** [comment]
Why is `Handyman 12` not listed before `Handyman 2`?

**Comment by Nubbie** [comment]
Guess it fixes: #1038 :+1:

**Comment by Broxzier** [comment]
I just noticed that the staff list will not be sorted if it is out of order already, because they don't switch names. To solve that the sort function will have to be rewritten to actually sort properly.

**Comment by IntelOrca** [comment]
> I just noticed that the staff list will not be sorted if it is out of order already, because they don't switch names. To solve that the sort function will have to be rewritten to actually sort properly.

I thought your trick when loading games fixed that.

**Comment by Broxzier** [comment]
I thought so too, but it doesn't work for staff, only guests.

**Comment by IntelOrca** [comment]
That is because `peep_update_names` only enumerates guests. You should make a new function to sort everyone by looking at how `peep_update_names` does it.

**Comment by IntelOrca** [comment]
@Broxzier To fix your issue you need to do a bit more work:

1. Count the total number of peeps. Use `FOR_ALL_PEEPS`.
2. malloc number of peeps * uint16 as peepList
3. Use `FOR_ALL_PEEPS` to set each peepList to the sprite index of each peep.
4. Do a qsort on peepList using your new comparator function.
5. Loop through the sorted peepList and update the linked list pointers, `peep->next` and `peep->previous`.
6. free peepList.

**Comment by Broxzier** [comment]
I'm not entirely sure how the peep list works. Are they just numbers? `peep_sort` is not finished, because I keep getting access violation errors after when calling `qsort`.

```
void peep_sort()
{
	uint16 sprite_index, num_peeps = 0;
	rct_peep *peep;
	FOR_ALL_PEEPS(sprite_index, peep) {
		num_peeps++;
	}

	uint16 *peep_list = (uint16*)malloc(num_peeps * sizeof(uint16));
	int i = 0;
	FOR_ALL_PEEPS(sprite_index, peep) {
		peep_list[i] = peep->sprite_index;
	}
	qsort(peep_list, num_peeps, sizeof(uint16), peep_compare);
	free(peep_list);
}
```

`peep_compare` now first checks if the two peeps have either "Guest X" as name, or a random generated one, and then simply compare the two numbers, only after that it would compare their actual names. But before this, it already triggers the error.

Is this not right: `rct_peep const *peep_a = GET_PEEP(*sprite_index_a);`?

**Comment by IntelOrca** [comment]
Use `sprite_index` instead of `peep->sprite_index`. I think the main problem you have is that you forgot to increment i in the second peep loop.

**Comment by Broxzier** [comment]
incrementing `i` made it work it so far. Now I'm fixing 'next' and 'previous', and after that should the list be copied over to the original list? `RCT2_ADDRESS_SPRITES_START_PEEP`?

**Comment by IntelOrca** [comment]
@Broxzier yeah, the actual sprite index of each peep can't change as there will be many things that reference them directly - like ride queues and vehicles. So that is you have this linked list as a way of enumerating all the peeps in a particular order.

**Comment by Broxzier** [comment]
The compare function now first checks peep types, which will keep all the guests and staff next to each other. Perhaps this can be used for a performance improvement, as you don't have to loop over all guests and check their types anymore.

The real names list has been alphabetically ordered. This means that the names will NOT be the same as in Vanilla RCT2, however I highly doubt anyone will notice. Most names are still the same, only a few had to be switched.

I moved the `peep_compare_name` function to utils.c, as it can be used for more than just peeps.

**Comment by Broxzier** [comment]
Rebased and ready to be tested.

File: src/peep/peep.c [hunk 1]
```diff
@@ -10389,46 +10410,73 @@ void peep_update_name_sort(rct_peep *peep)
...
+	// Make sure the first peep is set
+	gSpriteListHead[SPRITE_LIST_PEEP] = peep_list[0];
+
+	free(peep_list);
+
+	i = 0;
+	FOR_ALL_PEEPS(sprite_index, peep) {
+		i++;
+	}
+	assert(i == num_peeps);
```
Comment (IntelOrca) [inline src/peep/peep.c]: Perhaps wrap this last block in `#if DEBUG`?

Comment (Broxzier) [inline src/peep/peep.c]: Can do, but it's only done once when loading a park. In  case there is anything wrong, we'll find out much sooner.

**Comment by janisozaur** [comment]
```
/home/janisozaur/workspace/OpenRCT2/src/peep/peep.c: In function ‘peep_sort’:
/home/janisozaur/workspace/OpenRCT2/src/peep/peep.c:10432:46: warning: passing argument 4 of ‘qsort’ from incompatible pointer type [-Wincompatible-pointer-types]
  qsort(peep_list, num_peeps, sizeof(uint16), peep_compare);
                                              ^~~~~~~~~~~~
In file included from /home/janisozaur/workspace/OpenRCT2/src/peep/../rct2.h:34:0,
                 from /home/janisozaur/workspace/OpenRCT2/src/peep/../common.h:21,
                 from /home/janisozaur/workspace/OpenRCT2/src/peep/../addresses.h:20,
                 from /home/janisozaur/workspace/OpenRCT2/src/peep/peep.c:17:
/usr/include/stdlib.h:764:13: note: expected ‘__compar_fn_t {aka int (*)(const void *, const void *)}’ but argument is of type ‘int (*)(const uint16 *, const uint16 *) {aka int (*)(const short unsigned int *, const short unsigned int *)}’
 extern void qsort (void *__base, size_t __nmemb, size_t __size,
             ^~~~~
```

Other than that it seems to work.

**Comment by Broxzier** [comment]
Thanks. I didn't get the warning myself, but this commit should fix it.

**Comment by IntelOrca** [comment]
@Broxzier the warning was when building with GCC / Linux. You can see warnings from the CI by clicking *Show all checks* then details of AppVeyor and Travis.

**Comment by Gymnasiast** [comment]
Seems to work correctly. :+1:

**Comment by janisozaur** [comment]
I found it to be a cause of network desyncs. The results are non-deterministic on client and server, making `peep->next` diverge and lose sync later on.

**Comment by duncanspumpkin** [comment]
Yes this is guaranteed to cause a desync if the order call isn't sent to all clients.

**Comment by Broxzier** [comment]
I don't really see how. The only thing that really changes it the order of which peep logic is executed. When a new guest is spawned, it's inserted in the list at the right position, which should happen on all clients, and when someone joins the list will already be sorted, so their sort shouldn't change anything.

**Comment by duncanspumpkin** [comment]
Oh it looked to me like this ordering wouldn't happen on a new peep. If it does happen then yeah that's fine

**Comment by IntelOrca** [comment]
I was going to say that it re-orders the list on loading an SV6, but I don't think that happens for multiplayer.

**Pull request merged**